Add scripts to ease migration to new WLCG hierarchical scope structure

gregcorbett opened the pull request:
Resolves GT-1074

Add scripts should automate creation and then application of the new scope tags.

WLCGScopeTagCreateRunner.php, desired behaviour: Create every combination of VO.tierN for the supplied WLCG VO and tierN scope tags.

WLCGScopeTagAddRunner.php, desired behaviour: Sites get every possible combination of their existing WLCG VO and tierN scope tags in the new hierarchical structure.

I'd particularly appreciate checks that the logic is correct and achieves the desired behaviour.

gregcorbett committed:
Add scripts to ease migration to new WLCG heirachical scope structure

- scripts should create and then apply the new scope tags

tofu-rocketry commented on resources/wlcg-scope-tag-migration/WLCGScopeTagList.php:
```diff
@@ -0,0 +1,18 @@
+<?php
+
+/*
+ * A simple script to home some variables useful for the WLCG Scope tag
```
"Home" as in a place for them to live? "Store"?

gregcorbett replied:
"Store" is better 😄, addressed in https://github.com/GOCDB/gocdb/pull/548/commits/a271f3726521ac7766a45cf2b5d0675787ef2343.

tofu-rocketry commented on resources/wlcg-scope-tag-migration/WLCGScopeTagList.php:
```diff
@@ -0,0 +1,18 @@
+ */
+
+$wlcgScopesList = ["alice", "atlas", "belle", "cms", "dune", "lhcb"];
+$tierScopesList = ["tier0", "tier1", "tier2", "tier3"];
```
Is every tier (theoretically) possible for every VO?

gregcorbett replied:
Indeed.

garaimanoj commented on resources/wlcg-scope-tag-migration/WLCGScopeTagList.php:
```diff
@@ -0,0 +1,18 @@
+
+// Retrieve all scope objects from the DB and store as an dict, with the
+// name as the key, for later reference.
+$scopeDql = "SELECT s FROM Scope s";
```
Do you mean `scopeSql `

tofu-rocketry replied:
It's [DQL ](https://www.doctrine-project.org/projects/doctrine-orm/en/3.5/reference/dql-doctrine-query-language.html)rather than SQL, but I did raise a query if the comment should make that clearer!

tofu-rocketry commented on resources/wlcg-scope-tag-migration/WLCGScopeTagList.php:
```diff
@@ -0,0 +1,18 @@
+
+// Retrieve all scope objects from the DB and store as an dict, with the
+// name as the key, for later reference.
+$scopeDql = "SELECT s FROM Scope s";
```
Should the comment be any more specific that this is a DQL query?

gregcorbett replied:
addressed in https://github.com/GOCDB/gocdb/pull/548/commits/649dc143212cb46b12b2d8e3cd18c6d596cda4b9.

garaimanoj commented on resources/wlcg-scope-tag-migration/WLCGScopeTagAddRunner.php:
```diff
@@ -0,0 +1,89 @@
+
+    // Sites can support multiple CERN VOs - at any and multiple tiers.
+    // Check each combination in turn.
+    foreach ($wlcgScopesList as $wlcgScope) {
```
Adding
`
if (!in_array($wlcgScope, $siteScopes)) {
        continue;
    }
`
after the first foreach loop would prevent looping through the second loop in case $wlcgScope is not in $siteScopes.

gregcorbett replied:
addressed in https://github.com/GOCDB/gocdb/pull/548/commits/ef10b9e1326b6fb7269b6b3a89b5ed966c69581b.

tofu-rocketry reviewed (commented):
I think the logic looks ok. Have you tested it?

garaimanoj commented on resources/wlcg-scope-tag-migration/WLCGScopeTagAddRunner.php:
```diff
@@ -0,0 +1,89 @@
+                if (in_array($newScopeName, $siteScopes)) {
+                    echo "Skipping, site already has " . $newScopeName . "\n";
+                    continue;
+                } else {
```
The else block is unnecessary because the continue already exits the current iteration

gregcorbett replied:
Indeed, addressed in https://github.com/GOCDB/gocdb/pull/548/commits/ef10b9e1326b6fb7269b6b3a89b5ed966c69581b.

gregcorbett committed:
Improve code flow and remove unneeded conditional

- check WLCG scopes first and if not applied to the site, don't
  bother checking for tier scopes
- remove unneeded else

gregcorbett committed:
Update script level comment for clarity

- 'store' is a better word than 'home' here

gregcorbett committed:
Clarify that DQL, not SQL, is used

gregcorbett commented:
> I think the logic looks ok. Have you tested it?

Yep, both before and after the structural changes in https://github.com/GOCDB/gocdb/pull/548/commits/ef10b9e1326b6fb7269b6b3a89b5ed966c69581b.

garaimanoj reviewed (approved):
LGTM
